pkg/storegateway/bucket_streaming_readers.go

```diff
@@ -40,12 +41,19 @@ func NewSeriesChunksStreamReader(client storegatewaypb.StoreGateway_SeriesClient
 		queryLimiter:        queryLimiter,
 		stats:               stats,
 		log:                 log,
+		closeCalled:         make(chan struct{}),
 	}
 }
 
 // Close cleans up all resources associated with this SeriesChunksStreamReader.
 // This method should only be called if StartBuffering is not called.
 func (s *SeriesChunksStreamReader) Close() {
+	select {
```

Comment:
I think this method is racy now - previously we only called `Close` if `StartBuffering` was not called, but it's now possible for `Close` to be called from both `StartBuffering()` and `blocksStoreQuerier.selectSorted()` at the same time.

If this happens, both calls could check if `closeCalled` is open, see that it is still open, then both try to close the channel. One will succeed and the other will panic. I'm also not sure if it's safe to call `CloseSend()` twice or from two different goroutines at the same time.

(We should probably update the `This method should only be called if StartBuffering is not called.` comment above this method as well.)

Reply:
Alternatively, we could avoid this by only calling `StartStreaming()` after we know we want to use the stream (ie. after [here](https://github.com/grafana/mimir/blob/1720db083725d6128c972587c74340c137c19f1f/pkg/querier/blocks_store_queryable.go#L472-L476)) - this would maintain the symmetry with how ingester streaming behaves and remove the need for the `closeCalled` channel and associated race.

Reply:
Good idea! Thanks